File: src/engine/ai/calculations/PowerupCalculations.java
package engine.ai.calculations;

import java.util.ArrayList;

import engine.ai.controller.AiController;
import engine.entities.PhysicsObject;
import engine.entities.Player;
import engine.entities.PowerUp;
import engine.enums.ObjectType;
import engine.enums.PowerUpType;
import javafx.geometry.Point2D;
import javafx.scene.shape.Rectangle;

public class PowerupCalculations {

	//AI controller object
	protected AiController aiCon;
	//player object being controlled
	protected Player aiPlayer;
	//Map of game
	protected Rectangle map;
	//Array list of power ups
	protected ArrayList<PowerUp> powerups;
	//Player calculations object
	protected PlayersCalculations playerCalc;
	
	/**
	 * @param aiCon AI controller object
	 * @param map Map of game
	 * @param playerCalc Player calculations object
	 */
	public PowerupCalculations(AiController aiCon, Rectangle map, PlayersCalculations playerCalc) {

		this.aiCon = aiCon;
		this.aiPlayer = aiCon.getAiPlayer();
		this.map = map;
		this.playerCalc = playerCalc;
		powerups = new ArrayList<>();
		
	}
	
	/**
	 * Calculates location of neatest power up
	 * @return Index of nearest power up in 'powerups' array list, -1 if no power up is available
	 */
	public int getNearestPowerUp() {
		ArrayList<PowerUp> powerups = getPowerups();
		int index = -1;
		double distance = Double.MAX_VALUE;
		for (int i = 0; i < powerups.size(); i++) {
			double disToPU = calcDistance(powerups.get(i).getLocation(), aiPlayer.getLocation());
			if (disToPU < distance) {
				distance = disToPU;
				index = i;
			}
		}

		return index;
	}

	/**
	 * Calculates if a power up of given type exists in 'powerups' array list
	 * @param pu Power up type enquired about
	 * @return True if a power up of type given exists in 'powerups' array list, false otherwise
	 */
	public boolean powerUpExist(PowerUpType pu) {
		ArrayList<PowerUp> powerups = getPowerups();
		for (int i = 0; i < powerups.size(); i++) {
			if (powerups.get(i).getType().equals(pu)) {
				return true;
			}
		}
		return false;
	}
	
	/**
	 * Calculates location of power ups of type given
	 * @param pu Power up type enquired about
	 * @return Index of nearest power up of type given in 'powerups' array list, -1 if no power up of given type exist 
	 */
	public int getNearestPowerUp(PowerUpType pu) {
		ArrayList<PowerUp> powerups = getPowerups();
		int index = -1;
		double distance = Double.MAX_VALUE;
		for (int i = 0; i < powerups.size(); i++) {
			if (powerups.get(i).getType().equals(pu)) {
				double disToPU = calcDistance(powerups.get(i).getLocation(), aiPlayer.getLocation());
				if (disToPU < distance) {
					distance = disToPU;
					index = i;
				}
			}
		}

		return index;
	}
	
	/**
	 * @return Array list of all power ups
	 */
	public ArrayList<PowerUp> getPowerups(){
		updatePowerups();
		return powerups;
	}
	
	/**
	 * Calculates locations of nearest power up and player, and see which is closer
	 * @return True if nearest power up is closer than nearest player, false otherwise
	 */
	public boolean powerupCloserThanPlayer() {
		int puIndex = getNearestPowerUp();
		if(puIndex == -1)
			return false;
		double disToPu = powerups.get(puIndex).getLocation().distance(aiPlayer.getLocation());
		double disToPlayer = playerCalc.getNearestPlayer().getLocation().distance(aiPlayer.getLocation());
		
		return disToPu<disToPlayer;
	}
	
	/**
	 * Calculates distance between this AI player and locations of power ups available 
	 * @return True if distance to nearest power up is less than 0.2 of map's width, false otherwise
	 */
	public boolean powerupIsTooClose() {
		int index = getNearestPowerUp();
		if(index != -1 && isTooClose(powerups.get(index).getLocation())) {
			return true;
		}
		return false;
	}
	
	/**
	 * Updates 'powerups' array list from objects array list
	 */
	private void updatePowerups() {
		
		powerups.clear();
		ArrayList<PhysicsObject> objects = new ArrayList<>();
		objects.addAll(aiCon.getObjects());
		for (int i = 0; i < objects.size(); i++) {
			if (!objects.get(i).equals(null)) {
				if ( (objects.get(i).getTag()) ==  ObjectType.POWERUP)
					if (!powerups.contains(objects.get(i)))
						powerups.add((PowerUp) objects.get(i));
			}
		}
	}
	
	/**
	 * Calculates distance between AI player's location and given location
	 * @param loc Location to calculate distance to
	 * @return True distance between this AI player's location and given location is less than 0.2 of map's width, false otherwise
	 */
	private boolean isTooClose(Point2D loc ) {
		Point2D aiLoc = aiPlayer.getLocation();
		return (aiLoc.distance(loc)<(map.getWidth()*0.2));
	}
	
	/**
	 * Calculates distance between two points
	 * @param a The first point 
	 * @param b The second point
	 * @return distance between point a and b
	 */
	private double calcDistance(Point2D a, Point2D b) {
		return a.distance(b);
	}

}

File: neon-element-cpp/src/engine/ai/calculations/PowerupCalculations.cpp
#include "PowerupCalculations.hpp"
#include "engine/ai/controller/AiController.hpp"
#include <limits>

namespace neon {

PowerupCalculations::PowerupCalculations(AiController& aiCon, const Rect& map, PlayersCalculations& playerCalc)
    : aiCon_(aiCon), aiPlayer_(aiCon.getAiPlayer()), map_(map), playerCalc_(playerCalc)
{
}

int PowerupCalculations::getNearestPowerUp()
{
    auto& pups = getPowerups();
    int index = -1;
    double minDist = std::numeric_limits<double>::max();
    for (int i = 0; i < static_cast<int>(pups.size()); i++) {
        double d = static_cast<double>(distance(pups[static_cast<size_t>(i)]->getLocation(), aiPlayer_->getLocation()));
        if (d < minDist) {
            minDist = d;
            index = i;
        }
    }
    return index;
}

int PowerupCalculations::getNearestPowerUp(PowerUpType type)
{
    auto& pups = getPowerups();
    int index = -1;
    double minDist = std::numeric_limits<double>::max();
    for (int i = 0; i < static_cast<int>(pups.size()); i++) {
        if (pups[static_cast<size_t>(i)]->getType() == type) {
            double d = static_cast<double>(distance(pups[static_cast<size_t>(i)]->getLocation(), aiPlayer_->getLocation()));
            if (d < minDist) {
                minDist = d;
                index = i;
            }
        }
    }
    return index;
}

bool PowerupCalculations::powerUpExist(PowerUpType type)
{
    auto& pups = getPowerups();
    for (auto* p : pups) {
        if (p->getType() == type) return true;
    }
    return false;
}

std::vector<PowerUp*>& PowerupCalculations::getPowerups()
{
    updatePowerups();
    return powerups_;
}

bool PowerupCalculations::powerupCloserThanPlayer()
{
    int puIdx = getNearestPowerUp();
    if (puIdx == -1) return false;
    float distPu = distance(powerups_[static_cast<size_t>(puIdx)]->getLocation(), aiPlayer_->getLocation());
    Player* nearest = playerCalc_.getNearestPlayer();
    if (!nearest) return false;
    float distPlayer = distance(nearest->getLocation(), aiPlayer_->getLocation());
    return distPu < distPlayer;
}

bool PowerupCalculations::powerupIsTooClose()
{
    int idx = getNearestPowerUp();
    if (idx != -1 && isTooClose(powerups_[static_cast<size_t>(idx)]->getLocation()))
        return true;
    return false;
}

void PowerupCalculations::updatePowerups()
{
    powerups_.clear();
    auto& objects = aiCon_.getObjects();
    for (auto& obj : objects) {
        if (obj->getTag() == ObjectType::PowerUp) {
            powerups_.push_back(static_cast<PowerUp*>(obj.get()));
        }
    }
}

bool PowerupCalculations::isTooClose(Vec2 loc) const
{
    return distance(aiPlayer_->getLocation(), loc) < map_.width * 0.2f;
}

} // namespace neon
